cpp/src/preprocessing/quantize/detail/pq.cuh

```diff
@@ -0,0 +1,353 @@
+    auto stream = raft::resource::get_cuda_stream(res);
+    sub_labels_opt =
+      std::make_optional(rmm::device_uvector<uint32_t>(n_rows_train, stream, device_memory));
+    pq_cluster_sizes_opt =
+      std::make_optional(rmm::device_uvector<uint32_t>(pq_n_centers, stream, device_memory));
+    sub_labels_view_opt = std::make_optional(
+      raft::make_device_vector_view<uint32_t, ix_t>(sub_labels_opt.value().data(), n_rows_train));
+    pq_cluster_sizes_view_opt = std::make_optional(raft::make_device_vector_view<uint32_t, ix_t>(
+      pq_cluster_sizes_opt.value().data(), pq_n_centers));
+  }
+
+  for (ix_t m = 0; m < pq_dim; m++) {
+    RAFT_CUDA_TRY(cudaMemcpy2DAsync(sub_dataset.data_handle(),
+                                    sizeof(MathT) * pq_len,
+                                    trainset_ptr + m * pq_len,
+                                    sizeof(MathT) * dim,
+                                    sizeof(MathT) * pq_len,
+                                    n_rows_train,
+                                    cudaMemcpyDefault,
+                                    raft::resource::get_cuda_stream(res)));
+    auto pq_centers_subspace_view = raft::make_device_matrix_view<MathT, uint32_t, raft::row_major>(
+      pq_centers.data_handle() + m * pq_n_centers * pq_len, pq_n_centers, pq_len);
+    train_pq_centers<MathT, ix_t>(res,
+                                  params,
+                                  raft::make_const_mdspan(sub_dataset.view()),
+                                  pq_centers_subspace_view,
+                                  sub_labels_view_opt,
+                                  pq_cluster_sizes_view_opt);
+  }
+  return pq_centers;
+}
+
+template <typename T>
+quantizer<T> train(raft::resources const& res,
+                   const cuvs::preprocessing::quantize::pq::params params,
+                   raft::device_matrix_view<const T, int64_t> dataset)
+{
+  auto n_rows = dataset.extent(0);
+  auto dim    = dataset.extent(1);
+  raft::common::nvtx::range<cuvs::common::nvtx::domain::cuvs> fun_scope(
+    "preprocessing::quantize::pq::train(%zu, %u, %u, %u)",
+    size_t(n_rows),
+    dim,
+    params.pq_bits,
+    params.pq_dim);
+
+  auto ps = cuvs::neighbors::detail::fill_missing_params_heuristics(params, dataset);
+  std::optional<raft::device_matrix_view<const T, uint32_t, raft::row_major>>
+    vq_code_book_view_opt                                                           = std::nullopt;
+  std::optional<raft::device_matrix<T, uint32_t, raft::row_major>> vq_code_book_opt = std::nullopt;
+  if (ps.use_vq) {
+    vq_code_book_opt      = cuvs::neighbors::detail::train_vq<T>(res, ps, dataset);
+    vq_code_book_view_opt = raft::make_const_mdspan(vq_code_book_opt.value().view());
+  } else {
+    vq_code_book_opt = raft::make_device_matrix<T, uint32_t, raft::row_major>(res, 0, 0);
+  }
+  auto empty_codes = raft::make_device_matrix<uint8_t, int64_t, raft::row_major>(res, 0, 0);
+  if (ps.use_subspaces) {
```

Comment:
It would be great to use the PQ quantizer to do the quantization for IVF-PQ. Currently the output layout is different, but in the future we should unify these code paths. Could you create an issue to track that?